url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/GeomUtils.java

```diff
@@ -0,0 +1,1416 @@
+					// shall become:
+					// [ * * * j | a b L c d | F * * * * * ]
+					// |<-- active -->|<-- inactive -->|<-- forthcoming -->
+
+					int piBj = piB[j];
+					for (int pos = j; pos > activeBlength; --pos) {
+						piB[pos] = piB[pos - 1];
+					}
+					piB[activeBlength] = piBj;
+					++activeBlength;
+					--inactiveBlength;
+
+					int piBLast = piB[lastInactiveIndex];
+					for (int pos = lastInactiveIndex; pos > j; --pos) {
+						piB[pos] = piB[pos - 1];
+					}
+					piB[j] = piBLast;
+					systemHasChanged = true;
+
+					continue mainLoop;
+
+				}
+			}
+
+			// now there are only forthcoming active points.
+
+			// Now look for a point to add to the system:
+
+			// the "distance vector" from b to a:
+			double[] currentVec = {}; // LA-routines expand vector dimension
+			// when necessary.
+			for (int i = 0; i < inactiveAlength; ++i) {
+				currentVec = LA.plus(currentVec, LA.mul(+kktX[aStart + i],
+						a[piA[i]]));
+			}
+			for (int j = 0; j < inactiveBlength; ++j) {
+				currentVec = LA.plus(currentVec, LA.mul(-kktX[bStart - j],
+						b[piB[j]]));
+			}
+			result = LA.dot(currentVec, currentVec);
+
+			// if the distance is zero, the Lagrangians are both zero or the
+			// maximal inactive points are set,
+			// then the simplexes intersect.
+			if (scOp.equal(result, 0d) || scOp.equal(kktX[maxInactive + 0], 0d)
+					&& scOp.equal(kktX[maxInactive + 1], 0d) // the Lagrangians
+					// for a and b
+					|| inactiveAlength + inactiveBlength == maxInactive) {
+
+				break mainLoop;
+			}
+
+			// Now find the Lagrangian of the active point which has maximal
+			// value.
+			// Note that the sign of the Lagrangian DOES matter as this method
+			// computes the intersection of simplexes.
+			int iMax = -1;
+			double iMaxLagrangian = 0d;
+			for (int i = activeAlength + inactiveAlength; i < a.length; ++i) {
+				// difference to affine spaces: take sign into account (no
+				// Math.abs(...)).
+				double iLagrangian = -LA.dot(a[piA[i]], currentVec)
+						- kktX[maxInactive + 0];
+				if (iLagrangian > iMaxLagrangian) {
+					iMaxLagrangian = iLagrangian;
+					iMax = i;
+				}
+			}
+
+			int jMax = -1;
+			double jMaxLagrangian = 0d;
+			for (int j = activeBlength + inactiveBlength; j < b.length; ++j) {
+				// difference to affine spaces: take sign into account (no
+				// Math.abs(...)).
+				double jLagrangian = -LA.dot(b[piB[j]], currentVec)
+						- kktX[maxInactive + 1];
+				if (jLagrangian > jMaxLagrangian) {
+					jMaxLagrangian = jLagrangian;
+					jMax = j;
+				}
+			}
+
+			if (scOp.equal(iMaxLagrangian, 0d)
+					&& scOp.equal(jMaxLagrangian, 0d)) {
+
+				// then all forthcoming active points are perpendicular to the
+				// current vector
+
+				if (coeffsA != null) {
+					// return intersection location at a
+					Arrays.fill(coeffsA, 0d);
+					for (int i = activeAlength + inactiveAlength; i > activeAlength;) {
+						--i;
+						coeffsA[piA[i]] = kktX[aStart + i];
+					}
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 742:749, 750:759, 827:834, 835:843.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.